internal/reminder/reminder.go

```diff
@@ -149,70 +139,174 @@ func (r *reminder) Stop() {
 	})
 }
 
-// storeCursorState stores the cursor state to a file
-// Not thread-safe, should be called from a single goroutine
-func (r *reminder) storeCursorState(ctx context.Context) error {
+func (r *reminder) sendReminders(ctx context.Context) []error {
 	logger := zerolog.Ctx(ctx)
-	logger.Debug().Msg("storing cursor state")
 
-	var buf bytes.Buffer
-	enc := gob.NewEncoder(&buf)
+	// Fetch a batch of repositories
+	repos, err := r.getRepositoryBatch(ctx)
+	if err != nil {
+		logger.Error().Err(err).Msg("unable to fetch repositories")
+		return []error{err}
+	}
+
+	logger.Info().Msgf("created repository batch of size: %d", len(repos))
 
-	data := map[string]interface{}{
-		"projectListCursor": r.projectListCursor,
-		"repoListCursor":    r.repoListCursor,
+	// Update the reminder_last_sent for each repository to export as metrics
+	for _, repo := range repos {
+		logger.Debug().Msgf("updating reminder_last_sent for repository: %s", repo.ID)
+		err := r.store.UpdateReminderLastSentById(ctx, repo.ID)
+		if err != nil {
+			logger.Error().Err(err).Msgf("unable to update reminder_last_sent for repository: %s", repo.ID)
+			return []error{err}
+		}
+	}
+
+	// TODO: Send the actual reminders
+	return nil
+}
+
+func (r *reminder) getRepositoryBatch(ctx context.Context) ([]db.Repository, error) {
+	logger := zerolog.Ctx(ctx)
+	duration := pgtype.Interval{}
+	err := duration.Set(r.cfg.RecurrenceConfig.MinElapsed)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Debug().Msgf("fetching repositories after cursor: %s", r.repositoryCursor)
+	repos, err := r.store.ListEligibleRepositoriesAfterID(ctx, db.ListEligibleRepositoriesAfterIDParams{
+		ID:         r.repositoryCursor,
+		MinElapsed: duration,
+		Limit: sql.NullInt64{
+			Int64: int64(r.cfg.RecurrenceConfig.BatchSize),
+			Valid: true,
+		},
+	})
+	if err != nil {
+		return nil, err
 	}
-	if err := enc.Encode(data); err != nil {
-		return err
+
+	logger.Debug().Msgf("fetched %d repositories initially", len(repos))
+
+	intersectionPoint := -1
+	var additionalRepos []db.Repository
+
+	// Only fetch additional repositories if we are under the limit
```

Comment:
I don't think it's a requirement that we find _exactly_ `BatchSize` repos, just that we don't _exceed_ `BatchSize` repos in one pass.  We're trying to upper-bound the amount of background work added to the system.

I think this will get a lot simpler if we just do one pass:

```golang
repos, err := r.store.ListEligibleRepositoriesAfterID(ctx, {...})
if err != nil {
  return nil, err
}

// Don't actually ignore the error here...
if len(repos) == 0 || !RepositoryExistsAfterID(ctx, repos[len(repos)-1].ID) {
  r.repositoryCursor = uuid.UUID{}
} else {
  r.repositoryCursor = repos[len(repos)-1].ID
}

return repos, nil
```